operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/waitfor/ConditionChecker.java

```diff
@@ -0,0 +1,111 @@
+package io.javaoperatorsdk.operator.processing.dependent.waitfor;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.javaoperatorsdk.operator.api.reconciler.UpdateControl;
+import io.javaoperatorsdk.operator.api.reconciler.dependent.DependentResource;
+
+import static java.lang.Thread.sleep;
+
+public class ConditionChecker<R> {
+
+  public static final Duration DEFAULT_POLLING_INTERVAL = Duration.ofSeconds(1);
+  public static final Duration DEFAULT_TIMEOUT = Duration.ZERO;
+  private static final int MIN_THREAD_SLEEP = 25;
+
+  private Duration pollingInterval;
+  private Duration timeout;
+  private ConditionNotFulfilledHandler<?> conditionNotFulfilledHandler;
+  private Condition<R> condition;
+
+  public static <T> ConditionChecker<T> checker() {
+    return new ConditionChecker<>();
+  }
+
+  public ConditionChecker() {
+    this(DEFAULT_POLLING_INTERVAL, DEFAULT_TIMEOUT, UpdateControl::noUpdate);
+  }
+
+  public ConditionChecker(Duration pollingInterval, Duration timeout,
+      ConditionNotFulfilledHandler conditionNotFulfilledHandler) {
+    this.pollingInterval = pollingInterval;
+    this.timeout = timeout;
+    this.conditionNotFulfilledHandler = conditionNotFulfilledHandler;
+  }
+
+  public <P extends HasMetadata> void check(DependentResource<R, P> resource, P primary) {
+    check(() -> resource.getResource(primary));
+  }
+
+  public void check(Supplier<Optional<R>> supplier) {
+    checkSetup();
+    Optional<R> resource = supplier.get();
+    if (timeout.isNegative() || timeout.isZero()) {
+      if (resource.isPresent() && condition.isFulfilled(resource.get())) {
+        return;
+      } else {
+        handleConditionNotMet();
+      }
+    }
+    var deadline = Instant.now().plus(timeout.toMillis(), ChronoUnit.MILLIS);
+    while (Instant.now().isBefore(deadline)) {
+      resource = supplier.get();
+      if (resource.isPresent() && condition.isFulfilled(resource.get())) {
+        return;
+      } else {
+        var timeLeft = Duration.between(Instant.now(), deadline);
+        if (timeLeft.isZero() || timeLeft.isNegative()) {
+          handleConditionNotMet();
+        } else {
+          sleepUntilNextPoll(timeLeft);
+        }
+      }
+    }
+    handleConditionNotMet();
+  }
```

Comment:
Hi @scrocquesel , pls take a look here, that explains the use case for this:
https://github.com/java-operator-sdk/java-operator-sdk/issues/995#issuecomment-1060313476

TBH, I'm also kinda thinking to put it there, to explicitly support that use cases mentioned there. Since it's just for some small subset of operators. But as described, and alos shown in the integration tests, most condition check is async or that is perferred.

Reply:
> Also, the sleep loop is not resilient to an operator restart. I expect the timeout to last from the creation. If the operator stop just after the creation and restart after the timeout, then it should reconcile and immediately report a timeout and not wait for the timeout.

That would require a state management, what we deliberatily want to avoid if possible.

Reply:
Note that it's still a goal that on a reconciliation we reconcile every resource we manage. So if we want to have an analogy here this is not a state machine rather a workflow engine that is executed then on every reconciliation.